### ISSUE TITLE
ID and Breadcrumb JSON-LD collision

### ISSUE THREAD

**Issue by jono-alderson** [woocommerce/woocommerce#22440]
**Describe the bug**
Users running WooCommerce and Yoast SEO (and/or various other plugins which output/alter JSON-LD code) have invalid JSON-LD markup, due (in part) to the way in which WooCommerce outputs its markup.

**To Reproduce**
1. Install WooCommerce.
2. Install Yoast SEO, with breadcrumb markup activated.
3. Use Google's structured markup tester on a product page.
4. Observe the incorrect 'merging' of Breadcrumb and Product markup, where entities incorrectly share properties.

### Lots of moving parts here, so consolidating for simplicity.

- In JSON-LD markup, any entity can have an `ID` property (an arbitrary string). That ID uniquely identifies that entity, in the context of the page it's on.
- If multiple entities on a page share an ID, then search engines / etc may confuse and conflate those entities. In this case, the `product` and the `breadcrumbs` share the same ID, and Google is awkwardly trying to merge them.
- There's a special case with `breadcrumbs`, however, where the IDs _must_ be the URL of the page. So, no other entity on a page should ever use the URL as its ID. Note that, both Yoast and WooCommerce's JSON-LD handles this _correctly_.
- So, the `Product` needs a different, unique ID - and to _not_ use the URL of the page in question.
- For the sake of neatness, we recommend that an ID be structured as an 'extension' of the page the entity is on. E.g., in this case, the ID of the product could/should be `https://www.example.com/page/#product`, where the URL is the page in question.

**TL;DR**, please add `#product` to the end of the product ID string.

Making this change will remove the ID collision, and resolve the conflicts. Nothing else should change/break.

### A slight nuance
- If WooCommerce is referencing the _current_ ID anywhere (e.g., an `Organization` node may have a 'makesOffer' property which references the product ID), then the referenced IDs will also need to be updated to the new format.
- I don't _think_ that this is currently the case anywhere, but please check.

**Screenshots**
![1](https://user-images.githubusercontent.com/11629607/51105829-66715400-17ea-11e9-8a32-f9e04b0933f9.png)
![2](https://user-images.githubusercontent.com/11629607/51105835-696c4480-17ea-11e9-9df2-a816a9182f5a.png)

**Expected behavior**
The structured data testing tool should correclty represent the breadcrumbs and product(s) as separate entities, with the correct properties.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by rrennick** [comment]
Duplicate of #22045. If you still have a conflict with a SEO plugin after #22344 is applied then you should contact the SEO plugin author.

**Comment by jono-alderson** [comment]
This isn't a duplicate of #22045, and isn't fixed by #22344. The issue isn't the last breadcrumb member, it's that the ID of your `product` entity is incorrectly using the URL of the page as an ID. It should be using a unique value, such as `https://www.example.com/page/#product`.

**Comment by rrennick** [comment]
> Nor is it (in isolation) an SEO plugin conflict issue.

Can you provide steps to reproduce the issue with only WooCommerce active?

**Comment by jono-alderson** [comment]
No, because there are no other entities output by WooCommerce which use the page's URL as the entity ID. That doesn't make it other plugins' problem/fault - rather, you're essentially squatting on / using a global namespace and thus preventing other plugins from integrating.

Any other plugin/theme which uses Schema to output breadcrumbs (correctly) or any other schema (incorrectly, via using the page URL as the ID for any entity) is going to result in conflation.

This might be helpful reading: https://docs.google.com/document/d/14ITkylzQDQVOqM_pu8ZgzeKJD8VOJkPokrWo02FfElo/edit

**Comment by rrennick** [comment]
> A little confused now.

It was merged into master which means it will be in the next release of WC.

**Comment by jsmusiker** [comment]
@rrennick that I understand. I mean this issue from jono-alderson, is this the same as the original problem with the id merging? Or do you believe that #22344 has completely fixed the issue?

**Comment by rrennick** [comment]
I closed this issue because it is reporting a conflict with another plugin. The WordPress.org plugin repository has 54K+ plugins: https://wordpress.org/plugins/. 

It isn't practical to attempt to ensure that a plugin does not conflict with any of the other 54K plugins. In this case, the conflict is over the json+ld data for the product post type. The product post type is a feature of WC. If other plugins are going to output json+data for product content then those other plugins should either replace the WC data or output data that is compatible with the WC data.

**Comment by dfarin** [comment]
Awesome, searched about hours for words how i can explain this as an amateur and with terrible english, now its added for milestone. thx @ all that u take the time for reading, explaining and understanding this problem again.  :)

**Comment by claudiulodro** [comment]
Extending the URL with a hash value seems like a better solution than what was implemented in WC 3.5.4. 

Just to make sure we are on the same page, here is what I will do for WC 3.5.5:

- Revert the change introduced in #22344 so we include the product URL as the`@id` for the element in the Breadcrumblist.
- Append `#product` to the`@id` of the Product structured data, so the `@id` for the Product structured data will be e.g. `http://example.com/product/foo#product`.

If I've misunderstood the situation let me know. Thanks for the info @jono-alderson! 

Pull request incoming sometime this week.

**Comment by jono-alderson** [comment]
That's wonderful, thanks! 
Exciting!

**Comment by claudiulodro** [comment]
#22554
